state-chain/pallets/cf-ingress-egress/src/migrations/ccm_add_data_to_decoded.rs

```diff
@@ -0,0 +1,121 @@
+use cf_chains::{address::IntoForeignChainAddress, CcmChannelMetadataUnchecked};
+// Copyright 2025 Chainflip Labs GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+use frame_support::traits::UncheckedOnRuntimeUpgrade;
+
+use crate::Config;
+
+use crate::*;
+use frame_support::pallet_prelude::Weight;
+#[cfg(feature = "try-runtime")]
+use sp_runtime::DispatchError;
+
+use codec::{Decode, Encode};
+
+pub mod old {
+	use super::*;
+	use cf_chains::CcmAdditionalData;
+
+	#[derive(Clone, PartialEq, Eq, Encode, Decode)]
+	pub struct CrossChainMessage<C: Chain> {
+		pub egress_id: EgressId,
+		pub asset: C::ChainAsset,
+		pub amount: C::ChainAmount,
+		pub destination_address: C::ChainAccount,
+		pub message: CcmMessage,
+		pub source_chain: ForeignChain,
+		pub source_address: Option<ForeignChainAddress>,
+		pub ccm_additional_data: CcmAdditionalData,
+		pub gas_budget: GasAmount,
+	}
+
+	#[frame_support::storage_alias]
+	pub type ScheduledEgressCcm<T: Config<I>, I: 'static> = StorageValue<
+		Pallet<T, I>,
+		Vec<CrossChainMessage<<T as Config<I>>::TargetChain>>,
+		ValueQuery,
+	>;
+}
+
+pub struct CcmAdditionalDataToCheckedMigration<T: Config<I>, I: 'static>(PhantomData<(T, I)>);
+
+impl<T: Config<I>, I: 'static> UncheckedOnRuntimeUpgrade
+	for CcmAdditionalDataToCheckedMigration<T, I>
+{
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<Vec<u8>, DispatchError> {
+		let ccms = old::ScheduledEgressCcm::<T, I>::get()
+			.into_iter()
+			.map(|ccm| ccm.egress_id)
+			.collect::<BTreeSet<_>>();
+		Ok(ccms.encode())
+	}
+
+	fn on_runtime_upgrade() -> Weight {
+		log::info!("🍩 Running migration for Ingress-Egress pallet: Updating CCM's additional data to decoded.");
+		let _ = crate::ScheduledEgressCcm::<T, I>::translate::<
+			Vec<old::CrossChainMessage<<T as Config<I>>::TargetChain>>,
+			_,
+		>(|maybe_old_ccms| {
+			maybe_old_ccms.map(|old_ccms| {
+				old_ccms
+					.into_iter()
+					.map(|old_ccm| {
+						let checked_ccm = CcmChannelMetadataUnchecked {
+							message: old_ccm.message.clone(),
+							gas_budget: old_ccm.gas_budget,
+							ccm_additional_data: old_ccm.ccm_additional_data,
+						}
+						.to_checked(
+							old_ccm.asset.into(),
+							old_ccm.destination_address.clone().into_foreign_chain_address(),
+						)
+						.expect("All ccm should have valid additional data");
```

Comment:
Is this absolutely guaranteed? And if so, where? ... Think it's safer to just skip and log warning over any elements where it doesn't verify - since panicking here is very bad.

Reply:
I think it's guaranteed because we always (even before this PR) verified the CCM before storing it. Nonetheless I agree I think I'd rather we log an error and ignore instead of panicking.

Reply:
Ok changed to a filter_map instead and log an error when this fails